codepipeline-stack.yaml

```diff
@@ -0,0 +1,543 @@
+        "detail-type": ["CodePipeline Pipeline Execution State Change"],
+        "detail": {
+          "state": ["FAILED", "CANCELED", "SUCCEEDED"]
+        }
+      }
+      RoleArn: !GetAtt CodePipelineReadyFunctionRole.Arn
+      State: "ENABLED"
+      Targets: 
+        - Arn: !GetAtt CodePipelineReadyFunction.Arn
+          Id: "FailedPipeline"
+
+  CodePipelineReadyFunctionRole:
+    Type: AWS::IAM::Role
+    Properties:
+      AssumeRolePolicyDocument:
+        Statement:
+          - Action: sts:AssumeRole
+            Effect: Allow
+            Principal:
+              Service: lambda.amazonaws.com
+          - Action: sts:AssumeRole
+            Effect: Allow
+            Principal:
+              Service: codebuild.amazonaws.com  
+          - Action: sts:AssumeRole
+            Effect: Allow
+            Principal:
+              Service: events.amazonaws.com                           
+        Version: "2012-10-17"
+      ManagedPolicyArns:
+        - arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole
+        - arn:aws:iam::aws:policy/AWSCodePipelineCustomActionAccess
+  
+  CodePipelineReadyFunction:
+    Type: AWS::Lambda::Function
+    Properties:
+      Handler: index.handler
+      Role: !GetAtt CodePipelineReadyFunctionRole.Arn
+      Code:
+        ZipFile: |
+          import json
+          import os
+          import urllib3
+          import uuid
+          import logging
+
+          logger = logging.getLogger()
+          logger.setLevel(logging.DEBUG)
+
+          def handler(event, context):
+            logger.info('Received event{}'.format(event))
+
+            result = event['detail']['state']
+            status = "FAILURE"
+            
+            if result == 'SUCCEEDED':
+              status = 'SUCCESS'
+
+            encoded_body = json.dumps({
+              "Status": status,
+              "Reason": "CodePipeline Deploy ended",
+              "UniqueId": str(uuid.uuid4()),
+              "Data": "CodePipeline Deploy ended"
+            })
+
+            logger.info('Sending response {}'.format(encoded_body))
+            http = urllib3.PoolManager()
+            http.request('PUT', os.environ['SIGNAL_URL'], body=encoded_body)
+
+      Runtime: python3.9
+      Timeout: 900
+      Environment:
+        Variables:
+          SIGNAL_URL: !Ref CodePipelineWaitHandle
+
+  CodePipelineWaitHandle:
+    Type: AWS::CloudFormation::WaitConditionHandle
+
+  CodePipelineWaitCondition:
+    Type: AWS::CloudFormation::WaitCondition
+    Properties:
+      Handle: !Ref CodePipelineWaitHandle
+      Timeout: "5400"
+
+  PermissionForEventsToInvokeLambda: 
+    Type: AWS::Lambda::Permission
+    Properties: 
+      FunctionName: 
+        Ref: "CodePipelineReadyFunction"
+      Action: "lambda:InvokeFunction"
+      Principal: "events.amazonaws.com"
+      SourceArn: 
+        Fn::GetAtt: 
+          - "PipelineEventRule"
+          - "Arn"
+
+  PipelineSourceBucket:
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

S3 ObjectLockEnabled parameter not set to true. Make sure that you set S3 ObjectLockEnabled parameter to true. By setting the S3 ObjectLockEnabled parameter to true, you activate the Object Lock feature, providing added protection to your S3 objects by preventing deletion or overwriting for a specified retention period. This helps ensure data integrity, compliance, and regulatory requirements for your AWS S3 bucket.